### PR TITLE
Adds description for the form element manager to the documentation

### LINKED ISSUE #36: Documentation around `FormElementManager->get()`
This issue has been moved from the `zendframework` repository as part of the bug migration program as outlined here - http://framework.zend.com/blog/2016-04-11-issue-closures.html

---

Original Issue: https://api.github.com/repos/zendframework/zendframework/issues/7305
User: @claytondaley
Created On: 2015-03-10T15:31:30Z
Updated At: 2015-11-06T20:55:08Z
Body
I was using Custom Form elements in my application so I started to work through all the tips in the [Advanced Form](http://framework.zend.com/manual/current/en/modules/zend.form.advanced-use-of-forms.html) guide.   Despite implementing all of these points, I was still getting an error. Thanks to a reference example, I found that I could resolve my errors with one additional change not documented in the guide:

> You must also define the form (not just the custom elements) inside `form_elements` rather than `service_manager` otherwise `init()` is not run

If this is a requirement, it would be helpful to add it to the guide.  If this should not be a requirement, I may need to submit a bug report.

---

---

Originally posted by @GeeH at https://github.com/zendframework/zend-form/issues/92

### PR DESCRIPTION
**Pull request by froschdesign**

### CODE REVIEW

File: docs/book/v3/form-element-manager.md [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+## Create a Form Element Manager
+
+To create an instance, the form element manager requires a PSR-11 dependency container. The following examples uses the container implementation of [laminas-servicemanager](https://docs.laminas.dev/laminas-servicemanager/):
+
+```php
+$formElementManager = new Laminas\Form\FormElementManager(
+    new Laminas\ServiceManager\ServiceManager()
+);
+```
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md:36]: In practice, users will generally just pull the `FormElementManager` **_from_** the `ServiceManager`, though; this usage here is going to be quite rare, no?

File: docs/book/v3/form-element-manager.md [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+To get a supplied form element, use the class name of the element:
+
+```php
+$element = $formElementManager->get(Laminas\Form\Element\Select::class);
+```
+
+## Fetch and Configure a Standard Element
+
+Internally, the included factory `Laminas\Form\ElementFactory::class` for all types of elements, fieldsets and forms is used to create an element. This factory allows to configure an element, fieldset or form on creation:
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]: This feels passive to me. Maybe rephrase to:

```markdown
The form element manager uses the factory `Laminas\Form\ElementFactory` to create all elements, fieldsets, and forms.
```

and then remove the second sentence entirely (as it is redundant).

File: docs/book/v3/form-element-manager.md [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        'options' => [
+            'label'         => 'Rating',
+            'value_options' => [1, 2, 3, 4, 5],
+            ],
+        ],
+    ]
+);
+```
+
+The name for the element and the array with options are set as parameter to the constructor of the object on instantiation.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
The name for the element and the options array are provided as parameters to the associated class constructor on instantiation.
```

File: docs/book/v3/form-element-manager.md [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+```php
+$element->getName(); // rating
+$element->getLabel(); // Rating
+$element->getValueOptions(); // [1, 2, 3, 4, 5]
+```
+
+## Fetch a Custom Element without Registration
+
+The form element manager allows to fetch custom elements without prior registration on the manager.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
The form element manager allows fetching custom elements without prior registration with the manager.
```

File: docs/book/v3/form-element-manager.md [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+}
+```
+
+The form element manager can create these custom element by the related class name:
+
+```php
+$element = $formElementManager->get(ExampleElement::class);
+```
+
+Also here, the included factory `Laminas\Form\ElementFactory::class` is used by the manager to instance the element which allows the pass the name and options like before:
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
The manager uses the factory `Laminas\Form\ElementFactory` to instantiate the element, and will pass the name and options array just like in the prior example:
```

File: docs/book/v3/form-element-manager.md [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+```
+
+Register the custom element on the form element manager by using a factory class:
+
+```php
+$formElementManager->setFactory(
+    ExampleElement::class,
+    ExampleElementFactory::class
+);
+```
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md:139]: I wonder if we should demonstrate adding the factory _via configuration_ as well? And/or detail that programmatic addition like this should be done in a delegator factor for the `FormElementManager` service?

File: docs/book/v3/form-element-manager.md [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+}
+```
+
+The form element manager can create the form by the related class name:
+
+```php
+$form = $formElementManager->get(ExampleForm::class);
+```
+
+If no separate factory is required, then the form element manager will be instantiating the form class directly, by using the standard factory for elements `Laminas\Form\ElementFactory::class`.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
If no separate factory is required, then the form element manager will be instantiating the form class directly, by using the standard factory for elements (`Laminas\Form\ElementFactory`).
```

File: docs/book/v3/form-element-manager.md [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+```
+
+Register the form on the form element manager by using a factory class:
+
+```php
+$formElementManager->setFactory(
+    ExampleForm::class,
+    ExampleFormFactory::class
+);
+```
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md:201]: Same comment here as for lines 124-129.

File: docs/book/v3/form-element-manager.md [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+echo $form->getName(); // example
+```
+
+Now the custom factory will be used to instance the form.
+
+## Set a Paramater to a Form on Instantiation
+
+The options of a form can be used to set custom parameters to a form.
+Create a class for the form and get a custom option:
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]: Detail:

- You can hook into initialization via the `init()` method.
- The form element factory calls the `init()` method after creating an instance of the class.
- Since the name and options are processed in the constructor, you can access them via `init()` to perform further customizations for the instance.

File: docs/book/v3/form-element-manager.md [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+Fetch the form:
+
+```php
+$form = $formElementManager->get(ExampleForm::class);
+```
+
+The form element manager will provide the form with the custom element which is created by the form element manager, like before: with or without explicit registration of the element.
+
+
+## Why Using the Form Element Manager?
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
## Why Use the Form Element Manager?
```

File: docs/book/v3/form-element-manager.md [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+
+- Allows to overwrite or to extend elements without change form specifications.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
- It allows overwriting or extending elements without changing form specifications.
```

File: docs/book/v3/form-element-manager.md [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+- Allows decoration of elements, like adding a database adapter without passing the adapter through all layers of a form.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
- It allows decoration of elements, like adding a database adapter without passing the adapter through all layers of a form.
```

File: docs/book/v3/form-element-manager.md [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+- Handles correct instantiation with configuration and dependencies of elements, fieldsets and forms.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
- It manages instantiation and initialization of elements, fieldsets and forms, including all dependencies and configuration options.
```

File: docs/book/v3/form-element-manager.md [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+- Allows usage and configuration of custom elements without prior registration of the elements.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
- It allows usage and configuration of custom elements without prior registration of the elements.
```

File: docs/book/v3/form-element-manager.md [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+- Handles all types of objects of a form; elements, fieldsets and the form itself.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
- It handles all form object types: elements, fieldsets and the form itself.
```

**Review by weierophinney — CHANGES_REQUESTED** [review]

File: docs/book/v3/form-element-manager.md (hunk 6 above)
Comment (froschdesign) [reply docs/book/v3/form-element-manager.md:139]: > I wonder if we should demonstrate adding the factory _via configuration_ as well?

Right, but the configuration in an application is added on a separate page. I think this can be found better via the navigation and it doesn't make these pages even longer.

Comment (froschdesign) [reply docs/book/v3/form-element-manager.md:139]: I will check another option via tabs like we already have in laminas-i18n.

https://docs.laminas.dev/laminas-i18n/filters/alnum/#using-locale

Comment (weierophinney) [reply docs/book/v3/form-element-manager.md:139]: My main point is: in both laminas-mvc and Mezzio applications, you almost NEVER programmatically add factories to the form element manager. You would instead configure them.

The one place you might programmatically add them is via a delegator factory, though, again, that's specific to the Laminas ecosystem (as it's a feature of laminas-servicemanager).

But it comes back to: we should show the most common use cases as the primary examples, and then anything manual like this as a secondary example.

Comment (froschdesign) [reply docs/book/v3/form-element-manager.md:139]: _It's nice to see that the explanations for the integration in the application are now being taken into account. 😄_ 

---

To make it clear, the basics of the manager is to be explained here and nothing more, nothing less.
The usage in laminas-mvc and Mezzio are created separately. Then links and explicit coloured notes will also be added to this page here.
And the explanations for laminas-mvc and Mezzio will get their own entry in the navigation, because these explanations are important and should be clearly visible.

> …we should show the most common use cases as the primary examples, and then anything manual like this as a secondary example.

Unfortunately, this realization comes too late, because all documentations are written exactly the other way around. And now we have to deal with it.

See also the related project: https://github.com/orgs/laminas/projects/3

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+```php
+namespace Album\Controller;
+
+use Laminas\Form\FormElementManager;
+use Laminas\Mvc\Controller\AbstractActionController;
+
+final class AlbumController extends AbstractActionController
+{
+    public function __construct(
+        public readonly FormElementManager $formElementManager
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]: I'd recommend `private` here; we should not recommend `public` unless there's a really good reason.

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+    ) {}
+}
+```
+
+### Register Controller
+
+In a laminas-mvc based application, the form element manager is registered in the application during the [installation of laminas-form](../installation.md#installation-for-mezzio-and-laminas-mvc-application).
+This allows to fetch the form element manager from the application service container.
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
This allows fetching the form element manager from the application service container.
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+With the [reflection factory of laminas-servicemanager](https://docs.laminas.dev/laminas-servicemanager/reflection-abstract-factory/), the form element manager can be automaticly injected into the controller.
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
When using the [reflection factory of laminas-servicemanager](https://docs.laminas.dev/laminas-servicemanager/reflection-abstract-factory/), the form element manager can be automatically injected into the controller.
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+
+To [register the controller](https://docs.laminas.dev/laminas-mvc/quick-start/#create-a-route) for the application, extend the configuration of the module.
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
To [register the controller](https://docs.laminas.dev/laminas-mvc/quick-start/#create-a-route) in the application, extend the configuration of the module.
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+            Controller\AlbumController::class => ReflectionBasedAbstractFactory::class,
+        ],
+    ],
+    // …
+];
+</code></pre>
+
+## Fetch a Form without Registration
+
+The following example creates a class as form:
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
The following example creates a form class:
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 21]
```diff
@@ -0,0 +1,283 @@
...
+        // …
+    }
+}
+</code></pre>
+
+## Register and Fetch a Form With a Factory
+
+If a form needs some preparation for creation then a factory can be used.
+
+The following example creates a class as factory for the form:
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
The following example creates a factory class for the form:
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 22]
```diff
@@ -0,0 +1,283 @@
...
+        'factories' => [
+            // Add this line
+            Form\AlbumForm::class => Form\AlbumFormFactory::class,
+        ],
+    ],
+    // …
+];
+</code></pre>
+
+Get the form and the name in a controller:
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
To retrieve the form and the form name in a controller:
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+        assert($form instanceof AlbumForm);
+        
+        echo $form->getName(); // album
+        
+        // …
+    }
+}
+```
+
+Now the custom factory will be used to instance the form.
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
Now the custom factory will be used to create the form instance.
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 24]
```diff
@@ -0,0 +1,283 @@
...
+        assert($form instanceof AlbumForm);
+        
+        echo $form->get('example')->getLabel(); // Example element
+        
+        // …
+    }
+}
+```
+
+## Register and Using a Custom Element with a Factory
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
## Register and Use a Custom Element via a Factory
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 25]
```diff
@@ -0,0 +1,283 @@
...
+
+If a custom element needs some preparation for creation then a factory can be used.
+
+The following example creates a class as factory for the element of the previous example:
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
The following example creates a factory class for the element of the previous example:
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 26]
```diff
@@ -0,0 +1,283 @@
...
+        'factories' => [
+            // Add this line
+            Form\ExampleElement::class => Form\ExampleElementFactory::class,
+        ],
+    ],
+    // …
+];
+</code></pre>
+
+Now the custom factory will be used to instance the element.
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
Now the custom factory will be used to create the element instance.
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 27]
```diff
@@ -0,0 +1,283 @@
...
+
+## Configuring the Form Element Manager
+
+The [configuration of the form element manager follows the exact same pattern](https://docs.laminas.dev/laminas-servicemanager/configuring-the-service-manager/) as for a normal service manager of laminas-servicemanager.
+
+In a laminas-mvc based application this means to use the application or module configuration, like `config/autload/global.php` or `module/Album/config/module.config.php`, and the configuration key `form_elements`:
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
In a laminas-mvc based application this means using the application or module configuration, such as `config/autload/global.php` or `module/Album/config/module.config.php`, and the configuration key `form_elements`:
```

File: docs/book/v3/application-integration/form-element-manager-mvc.md [hunk 28]
```diff
@@ -0,0 +1,283 @@
...
+            'example' => Album\Form\ExampleElement::class,
+        ],
+        'abstract_factories' => [],
+        'delegators'         => [],
+        // …
+    ]
+];
+```
+
+The factory `Laminas\Form\FormElementManagerFactory` uses the configuration, search for the configuration key `form_elements` and create the form element manager.
```
Comment (weierophinney) [inline docs/book/v3/application-integration/form-element-manager-mvc.md]:
```suggestion
The factory `Laminas\Form\FormElementManagerFactory` uses the configuration, searches for the configuration key `form_elements` and creates the form element manager using the discovered configuration.
```

File: docs/book/v3/form-element-manager.md [hunk 29]
```diff
@@ -0,0 +1,318 @@
...
+
+## Benefits
+
+The following benefits are provided by the manager:
+
+- Handles all forms, fieldsets and elements, including all custom ones.
+- Create forms, fieldsets and elements with all dependencies like validators and filters.
+- Can create forms, fieldsets and elements with name and options.
+- Can provide elements, fieldsets and forms without prior registration.
+- Fetches hydrators and input filters from the application's service container and add them to a form or a fieldset.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
- Fetches hydrators and input filters from the application's service container and adds them to a form or a fieldset.
```

File: docs/book/v3/form-element-manager.md [hunk 30]
```diff
@@ -0,0 +1,318 @@
...
+- Allows to override existing elements or to extend them.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
- Allows overriding existing elements or extending them.
```

File: docs/book/v3/form-element-manager.md [hunk 31]
```diff
@@ -0,0 +1,318 @@
...
+
+> INFO: **Stand-Alone Usage**
+> The following examples show the basics of the form element manager using stand-alone usage.
+>
+> The configuration and usage in **laminas-mvc** or **Mezzio** based application [can be found in separate sections](#learn-more).
+
+## Create a Form Element Manager
+
+To create an instance, the form element manager requires a PSR-11 dependency container.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
To create an instance, the form element manager requires a [PSR-11 dependency container](https://www.php-fig.org/psr/psr-11/).
```

File: docs/book/v3/form-element-manager.md [hunk 32]
```diff
@@ -0,0 +1,318 @@
...
+The following examples uses the container implementation of [laminas-servicemanager](https://docs.laminas.dev/laminas-servicemanager/):
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
The following examples use the PSR-11 container implementation provided by [laminas-servicemanager](https://docs.laminas.dev/laminas-servicemanager/):
```

File: docs/book/v3/form-element-manager.md [hunk 33]
```diff
@@ -0,0 +1,318 @@
...
+The following example creates a custom element:
+
+```php
+final class ExampleElement extends Laminas\Form\Element
+{
+    // …
+}
+```
+
+The form element manager can create these custom element by the related class name:
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
The form element manager can create these custom elements by the related class name:
```

File: docs/book/v3/form-element-manager.md [hunk 34]
```diff
@@ -0,0 +1,318 @@
...
+        ],
+    ]
+);
+```
+
+## Register and Fetch a Custom Element with a Factory
+
+If a custom element needs some preparation for creation then a factory can be used.
+
+The following example creates a class as factory for the element:
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
The following example creates a factory class for the element:
```

File: docs/book/v3/form-element-manager.md [hunk 35]
```diff
@@ -0,0 +1,318 @@
...
+
+Get the custom element and the option:
+
+```php
+$element = $formElementManager->get(ExampleElement::class);
+
+echo $form->getOption('example_param'); // value
+```
+
+Now the custom factory will be used to instance the element.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
Now the custom factory will be used to create the element instance.
```

File: docs/book/v3/form-element-manager.md [hunk 36]
```diff
@@ -0,0 +1,318 @@
...
+
+## Fetch a Form without Registration
+
+The form element manager handles also whole forms.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
The form element manager also handles whole forms.
```

File: docs/book/v3/form-element-manager.md [hunk 37]
```diff
@@ -0,0 +1,318 @@
...
+
+The following example creates a class as form:
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
The following example creates a form class:
```

File: docs/book/v3/form-element-manager.md [hunk 38]
```diff
@@ -0,0 +1,318 @@
...
+
+Get the form and the name:
+
+```php
+$form = $formElementManager->get(ExampleForm::class);
+
+echo $form->getName(); // example
+```
+
+Now the custom factory will be used to instance the form.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
Now the custom factory will be used to instantiate the form.
```

File: docs/book/v3/form-element-manager.md [hunk 39]
```diff
@@ -0,0 +1,318 @@
...
+
+## Set a Paramater to a Form on Instantiation
+
+The options of a form can be used to set custom parameters to a form.
+
+* It can be hooked into the initialization via the `init()` method.
+* The form element factory calls the `init()` method after creating an instance of the class.
+* Since the name and options are processed in the constructor, it can be accessed them via `init()` to perform further customizations for the instance.
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]: These lines are bullet points... elsewhere, we've used `-` for bullet points, versus `*` as used here; what's the recommendation in our style guide?

Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
* It can be hooked into the initialization via the `init()` method.
* The form element factory calls the `init()` method after creating an instance of the class.
* Since the name and options are processed in the constructor, it can access them within `init()` to perform further customization of the instance.
```

File: docs/book/v3/form-element-manager.md [hunk 40]
```diff
@@ -0,0 +1,318 @@
...
+    {
+        /** @var mixed|null $exampleParam */
+        $exampleParam = $this->getOption('example_param');
+
+        // …
+    }
+}
+```
+
+The standard factory `Laminas\Form\ElementFactory` for form elements and forms is used which sets the name and/or the options to the object on creation via constructor:
```
Comment (weierophinney) [inline docs/book/v3/form-element-manager.md]:
```suggestion
The standard factory `Laminas\Form\ElementFactory` for form elements and forms is used which provides the name and/or the options to the object as arguments to the constructor:
```

**Review by weierophinney — CHANGES_REQUESTED** [review]
I've provided a number of suggested changes.

Additionally, I see a mixture of Oxford comma usage and non-Oxford comma usage; please standardize on one or the other (I'm not sure if we specify which style to use in the style guide...)

File: docs/book/v3/form-element-manager.md (hunk 39 above)
Comment (froschdesign) [reply docs/book/v3/form-element-manager.md]: Interesting that this is not recognised by the linter tool. :man_shrugging:

**Comment by froschdesign** [comment]
> Additionally, I see a mixture of Oxford comma usage and non-Oxford comma usage; please standardize on one or the other (I'm not sure if we specify which style to use in the style guide...)

This topic is not yet covered in the guide. I will add a note for it.
